iothub_client/src/iothub_client_core.c

```diff
@@ -1997,6 +2077,142 @@ IOTHUB_CLIENT_RESULT IoTHubClientCore_GetTwinAsync(IOTHUB_CLIENT_CORE_HANDLE iot
+            LogError("Could not start worker thread");
+        }
+        else
+        {
+            IOTHUB_QUEUE_CONSOLIDATED_CONTEXT* queueContext;
+
+            if ((queueContext = (IOTHUB_QUEUE_CONSOLIDATED_CONTEXT*)malloc(sizeof(IOTHUB_QUEUE_CONSOLIDATED_CONTEXT))) == NULL)
+            {
+                result = IOTHUB_CLIENT_ERROR;
+                LogError("Failed creating queue context");
+            }
+            else
+            {
+                queueContext->iotHubClientHandle = iotHubClientInstance;
+                queueContext->userCallback.getTwinSection = deviceTwinDesiredCallback;
+                queueContext->userContext = userContext;
+
+                // Codes_SRS_IOTHUBCLIENT_09_012: [ `IoTHubClientCore_GetTwinDesiredAsync` shall be made thread-safe by using the lock created in IoTHubClient_Create. ]
+                if (Lock(iotHubClientInstance->LockHandle) != LOCK_OK)
+                {
+                    // Codes_SRS_IOTHUBCLIENT_09_013: [ If acquiring the lock fails, `IoTHubClientCore_GetTwinDesiredAsync` shall return `IOTHUB_CLIENT_ERROR`. ]
+                    result = IOTHUB_CLIENT_ERROR;
+                    LogError("Could not acquire lock");
+                    free(queueContext);
+                }
+                else
+                {
+                    // Codes_SRS_IOTHUBCLIENT_09_014: [ `IoTHubClientCore_GetTwinDesiredAsync` shall call `IoTHubClientCore_LL_GetTwinDesiredAsync`, passing the `IoTHubClient_LL handle`, `deviceTwinCallback` and `userContext` as arguments ]
+                    // Codes_SRS_IOTHUBCLIENT_09_015: [ When `IoTHubClientCore_LL_GetTwinDesiredAsync` is called, `IoTHubClientCore_GetTwinDesiredAsync` shall return the result of `IoTHubClientCore_LL_GetTwinDesiredAsync`. ]
+                    result = IoTHubClientCore_LL_GetTwinDesiredAsync(iotHubClientInstance->IoTHubClientLLHandle, twinRequestOptions, iothub_ll_get_device_twin_section_async_callback, queueContext);
+
+                    if (result != IOTHUB_CLIENT_OK)
+                    {
+                        LogError("IoTHubClientCore_LL_GetTwinDesiredAsync failed");
+                        free(queueContext);
+                    }
+
+                    (void)Unlock(iotHubClientInstance->LockHandle);
+                }
+            }
+        }
+    }
+    return result;
+}
+
+IOTHUB_CLIENT_RESULT IoTHubClientCore_GetTwinReportedAsync(IOTHUB_CLIENT_CORE_HANDLE iotHubClientHandle, IOTHUB_CLIENT_GET_TWIN_REQUEST_OPTIONS* twinRequestOptions, IOTHUB_CLIENT_DEVICE_TWIN_SECTION_CALLBACK deviceTwinReportedCallback, void* userContext)
+{
+    IOTHUB_CLIENT_RESULT result;
+
+    // Codes_SRS_IOTHUBCLIENT_09_009: [ If `iotHubClientHandle` or `deviceTwinReportedCallback` are `NULL`, `IoTHubClientCore_GetTwinReportedAsync` shall return `IOTHUB_CLIENT_INVALID_ARG`. ]
+    if (iotHubClientHandle == NULL || deviceTwinReportedCallback == NULL)
```

Comment:
ugh, this code looks like pretty dup of the above function.

Reply:
Yeah. I tried to eliminate code duplication by using a common underlying function. But after reviewing with John and Ewerton we agreed better to do it like this.  To use common code would require passing an additional parameter so the correct "next layer" function would be called.  This parameter would likely need to be an enum in the common file, an enum at each layer, or exposing an existing enum from the MQTT transport layer.  Anyway, deemed not worth adding an enum or exposing the transport layer for this narrow of a purpose.